Adding Knative Eventing Object Model targeting 0.1 release. (#482)
knative/eventing: docs/spec/overview.md

multiple _Subscriptions_.

For more details, see [Kind: Source](spec.md#kind-source).

## Channel

**Channel** provides an at least once event delivery mechanism which can fan
out received events to multiple destinations via _Subscriptions_. A _Channel_
has a single inbound _Sinkable_ interface which may accept events from multiple
_Subscriptions_ or even direct delivery from external systems. Different
_Channels_ may implement different degrees of persistence. Event delivery order
is dependent on the backing implementation of the _Channel_ provided by the
_Provisioner_.

Event selection on a _Channel_ is 1:N – a single _Channel_ may fan out to
multiple _Subscriptions_.

See [Kind: Channel](spec.md#kind-channel).

## Provisioner

**Provisioner** catalogs available implementations of event _Sources_ and
_Channels_. _Provisioners_ hold a JSON Schema that is used to validate the
_Source_ and _Channel_ input parameters. _Provisioners_ make it possible to
provide cluster wide defaults for the _Sources_ and _Channels_ they provision.

_Provisioners_ do not directly handle events. They are 1:N with _Sources_ and
_Channels_.

> For more details, see [Kind: Provider](spec.md#kind-provisioner).

Reviewer: @sjwoodman  yep - I've sent a PR against the branch of @n3wscott for these typos:

https://github.com/n3wscott/eventing/pull/6

> For more details, see [Kind: Provider](spec.md#kind-provisioner).

Reviewer 2: 👍

# Overview

In the serverless model, computation and storage can be intertwined by having
storage and compute systems publish events which are handled asynchronously by
further compute services. Event delivery provides common infrastructure for
decoupling event producers and consumers such that the linking between them can
be performed without change to either. Loosely coupled eventing systems enable
composable services which retain the ability to scale independently.
Declarative event binding reduces the effort to create a scalable system that
combines managed services and custom code.

Knative eventing implements common components of an event delivery ecosystem:
enumeration and discovery of event sources, configuration and management of
event transport, and declarative binding of events (generated either by storage
services or earlier computation) to further event processing and persistence.

The following is a detailed specification for the `eventing.knative.dev` API

> surface, which provides control mechanism for Knative eventing.

Author: I think they are slightly different with different audiences, I have been assuming the motivation page is for people to find out _why_ we are doing this, and the overview page is for new devs to find out _what_ we are doing and the components at a high level. So if that is true, I think it is ok to have some duplication of the wording.

> surface, which provides control mechanism for Knative eventing.

Reviewer: Like @evankanderson I think it's more motivation - I came here to comment that the section is duplicated, IMO is best served in _Motivation_ ?

# Resource Types

The eventing API defines several resource types as well as interfaces which may
be fulfilled by other Kubernetes objects ([Sinkable](interfaces.md#sinkable),
[Subscribable](interfaces.md#subscribable),
[Channelable](interfaces.md#channelable),

> [Targetable](interfaces.md#targetable)). The API defines and provides a

Author: I moved them to the end of the Resource types section with a link to the interface page.

complete implementation for [Subscription](spec.md#kind-subscription), and
abstract resource definitions for [Sources](spec.md#kind-source),
[Channels](spec.md#kind-channel), and [Providers](spec.md#kind-provisioner)
which may be fulfilled by multiple backing implementations (much like the
Kubernetes Ingress resource).

 * A **Subscription** describes the transformation of an event (via a

>    _Callable_) and optional storage of a returned event. 

Reviewer 3: Agreed. Storage is an implementation detail of a sinkable. Maybe `s/storage/forwarding/`?

 * A **Source** represents an ongoing selection of events from an external
   system. A Subscription is used to connect these events to subsequent
   processing steps.

>  * A **Channel** provides event storage and fanout of events from a well-known

Author: Strangely this site does not seem to see the line end char as a char and wraps at n+1 the setting for line width.

>  * A **Channel** provides event storage and fanout of events from a well-known

Author: good points, I am using the word persistence. But I wonder if durability is more correct?

>  * A **Channel** provides event storage and fanout of events from a well-known

Reviewer 4: I prefer durability since it's a property of the event rather than the mechanism.

>    input address to multiple outputs described by Subscriptions. 

Reviewer 5: maybe bit of a nit here. but the way we describe these resources here are:
 - Describes, represents and provides
it seems a bit of mismatch. Channel says what it does, not what it represents. I think it would be better to try to rephrase all these as logical constructs in the system instead of what they do.
 Source: represents incoming events from an external system (and how to subscribe to them [subscribable])
 Subscription: represents an (optional) event transformation from a Source via Callable for (optionally) further processing via Channels [subscribable].
 Channel: Eventing middleware that can be subscribed to by 0 or more Subscriptions for processing.

Also, I'd like to see little clearer that Source / Channel are very symmetrical in the system from the control plane perspective. Source is Subscribable just like Channel. So Source shoves things into a Sinkable just like Subscription (possibly if so configured) shoves things into SInkable. The way I read this section, Source is somehow special from consumption part and it's really not.

Author: I gave this a shot but it might be something we have to do a follow up on with more detail when we work this model out a bit more.

<!-- This image is sourced from https://drive.google.com/open?id=10mmXzDb8S_4_ZG_hcBr7s4HPISyBqcqeJLTXLwkilRc -->
![Resource Types Overview](images/resource-types-overview.svg)

>  * **Provisioners** act as a catalog of the types of Sources and Channels

Author: updated to 
```
- **Provisioners** implement strategies for realizing backing resources for
  different implementations of _Sources_ and _Channels_ currently active in the
  eventing system.
```

   currently active in the eventing system.

<!-- This image is sourced from https://drive.google.com/open?id=1o_0Xh5VjwpQ7Px08h_Q4qnaOdMjt4yCEPixRFwJQjh8 -->

> ![Resource Types Provisioners](images/resource-types-provisioner.svg)

Author: I have left the channel-like object in the diagram and updated it to make it look like it implements the subscribable interface. Take a look. Reopen if this is not better :D

## Subscription

**Subscriptions** define the transport of events from one storage location

> (typically, **Source** or **Channel**) to the next (typically, a **Channel**)

Reviewer 3: A Subscription is not defining a transport; that's the Channel. And it should not be described as connecting a Channel to a Channel; that would be a "bridge" and just one relatively rare way of defining a Subscription with a no-op Callable. Rather it's registering a subscriber (Callable) to be called when events are available on the Channel. That is the primary role. The secondary role is to take the result of that call, if any, and forward it to another Channel which ACKs if it is able to assume responsibility for the event.

> (typically, **Source** or **Channel**) to the next (typically, a **Channel**)

Author: De-bolded. 

reworded to reflect what Mark has said. But I think your comment gets into the contract of the Targetable and Sinkable vs staying with the responsibility of the subscription.

through optional transformations (such as a Knative Service which processes
CloudEvents over HTTP). A **Subscription** resolves the addresses of
transformations (`call`) and destination storage (`result`) through the
_Callable_ and _Sinkable_ interface contracts, and writes the resolved

> addresses to the _Subscribable_ `from` resource. A **Subscriptions** do not

Reviewer 6: A Subscriptions -> Subscriptions

to specify both a transformation and a storage destination, but at least one
must be provided.

> All event transport over a **Subscription** is 1:1 – only a single `from`,

Author: better?
```
All event delivery linkage from a **Subscription** is 1:1 – only a single
`from`, `call`, and `result` may be provided.
```

`call`, and `result` may be provided.

For more details, see [Kind: Subscription](spec.md#kind-subscription).

## Source

> **Source** represents an ongoing selection of events from an external system,

Author: those are all external to the application though so I think it still works.

such as object creation events in a specific storage bucket, database updates
in a particular table, or Kubernetes resource state changes. Because a
**Source** represents an external system, it only produces events (and is
therefore _Subscribable_ by **Subscriptions**). **Source** configures an
external system and may include parameters such as specific resource names,
event types, or credentials which should be used to establish the connection to
the external system. The set of allowed configuration parameters is described
by the **Provisioner** which is referenced by the **Source**.

Event selection on a **Source** is 1:N – a single **Source** may fan out to
multiple **Subscriptions**.

For more details, see [Kind: Source](spec.md#kind-source).

## Channel

**Channel** provides a durable event storage mechanism which can fan out
received events to multiple destinations via **Subscriptions**. A **Channel**
has a single inbound _Sinkable_ interface which may accept events from multiple
**Subscriptions** or even direct delivery from external systems. Different

> **Channels** may implement different degrees of persistence.

Reviewer 3: If "different degrees of persistence" includes the in-memory (stub) channel provisioner, "provides a durable event storage mechanism" would not be a defining characteristic of Channel.

> **Channels** may implement different degrees of persistence.

Author: good points. word tweaking

> **Channels** may implement different degrees of persistence.

Author: did it.

See [Kind: Channel](spec.md#kind-channel).

## Provisioner

The eventing API has parallel constructs for event _sources_ (systems which
create events based on internal or external changes) and event _transports_
(middleware which add value to the event delivery chain, such as persistence or
buffering).

For more details, see [Kind: Provider](spec.md#kind-provisioner).

--- 

> *Next*:

Author: I think having it makes the docs way more readable. Gonna leave it until we get real navigation, but I like the idea of unlinking the current page!

combines managed services and custom code.

Knative eventing implements common components of an event delivery ecosystem:
enumeration and discovery of event sources, configuration and management of
event transport, and declarative binding of events (generated either by storage
services or earlier computation) for further event processing and persistence.

The following is a detailed specification for the `eventing.knative.dev` API
surface, which provides control mechanism for Knative eventing.

# Resource Types

The API defines and provides a complete implementation for
[Subscription](spec.md#kind-subscription), and abstract resource definitions
for [Sources](spec.md#kind-source), [Channels](spec.md#kind-channel), and
[Providers](spec.md#kind-provisioner) which may be fulfilled by multiple
backing implementations (much like the Kubernetes Ingress resource).

- A **Subscription** describes the transformation of an event and optional
  forwarding of a returned event.

- A **Source** allows an incoming events from an external system to be
  _Subscribable_. A _Subscription_ is used to connect these events to
  subsequent processing steps.

> - A **Channel** provides event persistance and fanout of events from a

Reviewer: Persistence is than defined by the actual implementation (e.g. Kafka topics) ?

  well-known input address to multiple outputs described by _Subscriptions_.

<!-- This image is sourced from https://drive.google.com/open?id=10mmXzDb8S_4_ZG_hcBr7s4HPISyBqcqeJLTXLwkilRc -->

![Resource Types Overview](images/resource-types-overview.svg)

- **Provisioners** implement strategies for realizing backing resources for
  different implementations of _Sources_ and _Channels_ currently active in the
  eventing system.

<!-- This image is sourced from https://drive.google.com/open?id=1o_0Xh5VjwpQ7Px08h_Q4qnaOdMjt4yCEPixRFwJQjh8 -->

![Resource Types Provisioners](images/resource-types-provisioner.svg)

With extendibility and compostability as a goal of Knative Eventing, the
eventing API defines several resources that can be reduced down to a well
understood contracts. These eventing resource interfaces may be fulfilled by
other Kubernetes objects and then composed in the same way as the concreate
objects. The interfaces are ([Sinkable](interfaces.md#sinkable),
[Subscribable](interfaces.md#subscribable),
[Channelable](interfaces.md#channelable),
[Targetable](interfaces.md#targetable)). For more details, see
[Interface Contracts](interfaces.md).

## Subscription

> **Subscriptions** describe a flow of events from one event producer or

Reviewer: * Each `Subscription` is basically tied to two "components" (e.g. event source sends events to some target).
* Now that _target_ could process the data, act as a _source_ and send it to a different target
* and so on...

All of these `Subscritptions` may represent some "Workflow".  Are there plans to cover this too? (Perhaps the the via the "CNCF Serverless WG's `workflow` spec)  Just wondering - I do *NOT* think that's in scope for 0.1 - but just curious

Reviewer 7: I think there are interesting questions about how to chain Targetables together.

For the scope of 0.1, the only option we support is sinking the result of one Targetable into a Channel, and then creating a Subscription from that Channel which calls the next Targetable.

Author: Hm. This is interesting. Let me shop the idea around a bit and get back to you.

Reviewer 8: > Now that target could process the data, act as a source and send it to a different target

@matzew If I understand what you're proposing, this is a common anti-pattern in serverless. Each nested call ends up causing the original call to block preventing the component from scaling down once it had completed its work. 

![nested calls](https://user-images.githubusercontent.com/302992/46741297-b9c78980-cc72-11e8-9d66-6f8bf5355a5c.png)

The deeper the chain, the more components are blocked waiting for a reply, doing no real work, but generating billable time.

It's possible to avoid the blocking, but then you lose any ability to detect and recover from errors.

![linear calls](https://user-images.githubusercontent.com/302992/46741529-36f2fe80-cc73-11e8-81a3-4990d2087a07.png)

The current Subscription model creates a chain or responsibility where channels mediate the events creating both error boundaries and the ability to retry after failures.

![chain of responsibility-2](https://user-images.githubusercontent.com/302992/46741628-70c40500-cc73-11e8-9768-e858763c7767.png)

Treat any references to "Call Pipe" as a "Channel" these diagrams were created against a much, much earlier proposal.

Reviewer: > Now that target could process the data, act as a source and send it to a different target

@scothis  one more comment/thought.

So, the `target` (aka function, aka `ksvc`) receives data. But it would not do a direct (http) invocation. I think the function here could act as a _true_ event-source: Delivering an _async_ event, after it (the `target`) did process the data.

Now, this _event-source_ (implemented as a function) can have subscriptions again ... 

@n3wscott @evankanderson  any thoughts ?

Reviewer: Ok... I think my comments are _irrelevant_ for *this* PR - I will open an issue for discussions

forwarder (typically, _Source_ or _Channel_) to the next (typically, a
_Channel_) through transformations (such as a Knative Service which processes
CloudEvents over HTTP). A _Subscription_ controller resolves the addresses of
transformations (`call`) and destination storage (`result`) through the
_Targetable_ and _Sinkable_ interface contracts, and writes the resolved
addresses to the _Subscribable_ `from` resource. _Subscriptions_ do not need to
specify both a transformation and a storage destination, but at least one must
be provided.

All event delivery linkage from a **Subscription** is 1:1 – only a single
`from`, `call`, and `result` may be provided.

> For more details, see [Kind: Subscription](spec.md#kind-subscription).

Reviewer 2: Does this imply that only a linear pipeline of function calls can be supported (regardless of the level of abstraction of deployment units)? Even if I manually deploy a set of _Subscription_, I could not introduce any conditional logic or branching?

Reviewer 9: That seems so to my eyes. I think the most important "building block" is missing in the spec: the spec is missing a *"standard contract for a service to publish events on a bus channel, regardless of the bus implementation"*.

"Callable" services are *not* necessarily functions and moreover they are not necessarily *pure functions* without side effects. I think a Service should be able to publish one, two or several events within the context of a invocation, provided that it has been given the permissions (e.g. via static resource configuration) to do so.

Reviewer 2: I agree, that I would expect a Service to be able to publish a number of events. This would also be convenient for error handling and enabling a different component to respond to errors within a Service.

Author: This has come up a couple times now. I think the correct answer is another spec that outlines the runtime contract spec for how subscribables, targetables and sinkables communicate at a protocol level and then how that maps to transports we support.

Reviewer 7: Thanks for calling out this ambiguity. The intent is that _Sinkable_ would be the interface for delivering an event (including to a function) which takes ownership of the event from a "chain of responsibility" pattern.

I think there is a second document we need that describes the data plane transport contracts. For now, our baseline is HTTP POST of CloudEvents, but a protocol-negotiation scheme would be very interesting once we've shipped sufficient primitives.

# Resource Types

The API defines and provides a complete implementation for
[Subscription](spec.md#kind-subscription), and abstract resource definitions

> for [Channels](spec.md#kind-channel) and [Providers](spec.md#kind-provisioner)

Reviewer 8: ```suggestion
for [Channels](spec.md#kind-channel) and [ClusterProvsioner](spec.md#kind-clusterprovisioner)
```

which may be fulfilled by multiple backing implementations (much like the
Kubernetes Ingress resource).

- A **Subscription** describes the transformation of an event and optional
  forwarding of a returned event.

- A **Channel** provides event persistance and fanout of events from a
  well-known input address to multiple outputs described by _Subscriptions_.

<!-- This image is sourced from https://drive.google.com/open?id=10mmXzDb8S_4_ZG_hcBr7s4HPISyBqcqeJLTXLwkilRc -->

![Resource Types Overview](images/resource-types-overview.svg)

- **ClusterProvisioners** implement strategies for realizing backing resources
  for different implementations of _Channels_ currently active in the eventing
  system.

<!-- This image is sourced from https://drive.google.com/open?id=1o_0Xh5VjwpQ7Px08h_Q4qnaOdMjt4yCEPixRFwJQjh8 -->

<img alt="Resource Types ClusterProvisioners" src="images/resource-types-provisioner.svg" width="200">

> With extendibility and compostability as a goal of Knative Eventing, the

Author: added `Sources are defined by independent CRDs that can be installed into a cluster.
For more information see [Knative Eventing
Sources](https://github.com/knative/eventing-sources).`

eventing API defines several resources that can be reduced down to a well
understood contracts. These eventing resource interfaces may be fulfilled by
other Kubernetes objects and then composed in the same way as the concreate
objects. The interfaces are ([Sinkable](interfaces.md#sinkable),
[Subscribable](interfaces.md#Subscribable),
[Targetable](interfaces.md#targetable)). For more details, see
[Interface Contracts](interfaces.md).

## Subscription

> **Subscriptions** describe a flow of events from one _Channel_) to the next

Author: rejected unless you pay me 12 bitcoin.

(this is a far reaching change that is more than this one update, it should be a follow-up because this style is used throughout the document.)

> **Subscriptions** describe a flow of events from one _Channel_) to the next

Reviewer 7: Will do.

# Resource Types

The API defines and provides a complete implementation for
[Subscription](spec.md#kind-subscription), and abstract resource definitions
for [Channels](spec.md#kind-channel) and

> [ClusterChannelProvisioners](spec.md#kind-cluster-channe-provisioner) which may

Reviewer 8: ```suggestion
[ClusterChannelProvisioners](spec.md#kind-clusterchannelprovisioner) which may
```

  system.

<!-- This image is sourced from https://drive.google.com/open?id=1o_0Xh5VjwpQ7Px08h_Q4qnaOdMjt4yCEPixRFwJQjh8 -->

<img alt="Resource Types ClusterChannelProvisioners" src="images/resource-types-provisioner.svg" width="200">

Sources are defined by independent CRDs that can be installed into a cluster.
For more information see [Knative Eventing
Sources](https://github.com/knative/eventing-sources).

## Subscription

**Subscriptions** describe a flow of events from one _Channel_) to the next
Channel\* through transformations (such as a Knative Service which processes
CloudEvents over HTTP). A _Subscription_ controller resolves the addresses of
transformations (`call`) and destination storage (`result`) through the
_Targetable_ and _Sinkable_ interface contracts, and writes the resolved
addresses to the _Channel_ in the `from` reference. _Subscriptions_ do not need
to specify both a transformation and a storage destination, but at least one
must be provided.

All event delivery linkage from a **Subscription** is 1:1 – only a single
`from`, `call`, and `result` may be provided.

For more details, see [Kind: Subscription](spec.md#kind-subscription).

## Channel

> **Channel** provides an at least once event delivery mechanism which can fan

Reviewer 3: We shouldn't be providing any delivery guarantees. That will be dependent on how the logical Channel is backed via its Provisioner. For example, an in-memory Channel implementation that fails between receiving and delivering an event would not conform to an "at least once" guarantee.

out received events to multiple destinations via _Subscriptions_. A _Channel_
has a single inbound _Sinkable_ interface which may accept events from multiple
_Subscriptions_ or even direct delivery from external systems. Different
_Channels_ may implement different degrees of persistence. Event delivery order
is dependent on the backing implementation of the _Channel_ provided by the
_ClusterChannelProvisioner_.

Event selection on a _Channel_ is 1:N – a single _Channel_ may fan out to
multiple _Subscriptions_.

See [Kind: Channel](spec.md#kind-channel).

## ClusterChannelProvisioner

**ClusterChannelProvisioner** catalogs available implementations of _Channels_.
_ClusterChannelProvisioners_ hold a JSON Schema that is used to validate the
_Channel_ input arguments. _ClusterChannelProvisioners_ make it possible to
provide cluster wide defaults for the _Channels_ they provision.

_ClusterChannelProvisioners_ do not directly handle events. They are 1:N with
_Channels_.

For more details, see [Kind:

> ClusterChannelProvisioner](spec.md#kind-cluster-channel-provisioner).

Reviewer 8: ```suggestion
ClusterChannelProvisioner](spec.md#kind-clusterchannelprovisioner).
```
